Add files via upload

abi-awosola opened the pull request.

abi-awosola committed:
Add files via upload

larskotthoff reviewed (changes requested):
- How did you set up the nested resampling?
- What are the hyperparameters that you tuned?
- What are the steps of your ML pipeline?
- I didn't understand how the classification and regression parts fit together. Or are they completely separate? Please explain in detail what you're doing there.

abi-awosola commented:
[COSC_5557_PipelineOptmzn.pdf](https://github.com/COSC5557/pipeline-optimization-abi-awosola/files/15055201/COSC_5557_PipelineOptmzn.pdf)



How did you set up the nested resampling? Similar to my comment on HPO, I didn't use nested resampling. However, I adjusted my work to include cross validation with Optuna. I have more details in the file.

What are the hyperparameters that you tuned? I have included this in my adjustment.

What are the steps of your ML pipeline? This is my file also.

I didn't understand how the classification and regression parts fit together. Or are they completely separate? Please explain in detail what you're doing there.

I used different datasets. I have included this in the file.

abi-awosola commented:
I did not use nested resampling. Optuna does not need to use it. Do I have to use nested resampling?

The different number of trials are for different problems- classification and regression.

I can tune more than one hyperparameter per learner.

larskotthoff commented:
Yes, you have to use nested resampling.

abi-awosola commented:
[COSC_5557_PipelineOptmzn.pdf](https://github.com/COSC5557/pipeline-optimization-abi-awosola/files/15208290/COSC_5557_PipelineOptmzn.pdf)


Professor,

I will include the code in in GitHub.

Thank you.

larskotthoff commented:
- What hyperparameters of the preprocessing method did you tune?
- Not sure if I'm following how you determined the best hyperparameter configuration; can you describe in more detail please? In particular, you cannot just choose the configuration that had the best performance in the outer resampling. See our discussion in class of this issue.

abi-awosola commented:
- The `scaler` and the `selector threshold` were the preprocessing hyperparameters that were tuned.
- Oh, I see. The inner loop is used for hyperparameter tuning while the outer loop is used for model evaluation.

* From class notes:
- ( 'The inner evaluation is to find what works best based on the training data- the hyperparameter, while the outer evaluation on data not used in inner evaluation to get unbiased estimates of expected performance on new data' )

Thank you for the feedback, Professor. This is the idea I was trying to describe.

larskotthoff commented:
Can you update your report accordingly please?

abi-awosola commented:
[COSC_5557_PipelineOptmzn.pdf](https://github.com/COSC5557/pipeline-optimization-abi-awosola/files/15215640/COSC_5557_PipelineOptmzn.pdf)


Yes, here is the file.

larskotthoff commented:
Still not sure how you got the best hyperparameter settings -- you get several from your outer CV, right? How do you choose one?

abi-awosola commented:
[COSC_5557_PipelineOptmzn.pdf](https://github.com/COSC5557/pipeline-optimization-abi-awosola/files/15226391/COSC_5557_PipelineOptmzn.pdf)

**Selecting Best Hyperparameters**: After tuning, the best hyperparameters are those that yield the highest average performance across all folds in the outer cross-validation loop. I have included this details in the uploaded file.

larskotthoff commented:
You cannot do it like this, as we discussed in class.

abi-awosola commented:
I think aggregation was left out in my report. This is towards taking average at some point to provide unbiased performance estimate.


**Aggregation:**

* After completing the inner cross-validation loop for each configuration of hyperparameters, the average performance score across the inner folds is calculated.
* This average score is then appended to the `nested_scores` list.
* Similarly, once all outer fold iterations for a specific set of hyperparameters is done, the average performance score across all outer folds is computed by taking the mean of the scores in the `nested_scores` list.
*  The set of hyperparameters with the highest aggregated performance score is considered the optimal choice for deployment.

larskotthoff commented:
Again, we discussed this in class. You cannot do it like this. Just like for ML without hyperparameter optimization, the only purpose of the (nested) resampling is to provide unbiased performance estimates -- you do not get a model or hyperparameter setting from this. To get the model, you run on the entire dataset. To get the hyperparameter setting, you also run HPO on the entire dataset (i.e. without the outer resampling).

abi-awosola commented:
Hmm, I see.

On Mon, May 6, 2024, 16:59 Lars Kotthoff ***@***.***> wrote:

> Again, we discussed this in class. You cannot do it like this. Just like
> for ML without hyperparameter optimization, the only purpose of the
> (nested) resampling is to provide unbiased performance estimates -- you do
> not get a model or hyperparameter setting from this. To get the model, you
> run on the entire dataset. To get the hyperparameter setting, you also run
> HPO on the entire dataset (i.e. without the outer resampling).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/COSC5557/pipeline-optimization-abi-awosola/pull/2#issuecomment-2097065681>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BEAXPKK2QRET4WHN4JOVMXLZBADMHAVCNFSM6AAAAABGN4WOPWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAOJXGA3DKNRYGE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

abi-awosola commented:
[COSC_5557_PipelineOptmzn.pdf](https://github.com/COSC5557/pipeline-optimization-abi-awosola/files/15228660/COSC_5557_PipelineOptmzn.pdf)


Here, I have made adjustments based on the guidelines, Professor.

larskotthoff commented:
Don't understand your resampling -- you say you do a train/test split, then a 5 fold CV, and then a 3 fold CV?

abi-awosola commented:
The train/test split part was used to get performance before optimization.

On Tue, May 7, 2024 at 9:21 AM Lars Kotthoff ***@***.***>
wrote:

> Don't understand your resampling -- you say you do a train/test split,
> then a 5 fold CV, and then a 3 fold CV?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/COSC5557/pipeline-optimization-abi-awosola/pull/2#issuecomment-2098693846>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BEAXPKMVSIX72EJIKD3B4STZBDWN3AVCNFSM6AAAAABGN4WOPWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAOJYGY4TGOBUGY>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

larskotthoff commented:
This is unclear in your report. In particular you mention multiple different CVs and it's unclear how they work together.

abi-awosola commented:
[COSC_5557_PipelineOptmzn.pdf](https://github.com/COSC5557/pipeline-optimization-abi-awosola/files/15242058/COSC_5557_PipelineOptmzn.pdf)

You're right, Professor. I have included it now.

larskotthoff commented:
Quoting from your report:
> The training set is further partitioned into 5-fold/subset cross-validation.
> Each fold serves as a separate training-validation split for the outer cross-validation loop.

So you have a train/test split. Then you partition that into 5 folds. Then you partition each fold again into a train/test split? And that wouldn't be the outer resampling.

> Nested within the outer loop, the inner loop facilitates hyperparameter optimization by further
partitioning the training folds into 3 non-overlapping folds.

3 folds? Is that in addition to the 5 above? Or should the 3 be a 5? Please check your report carefully.

abi-awosola commented:
Thank you for the feedback, Professor.

I see the distinction.

On Tue, May 7, 2024, 19:34 Lars Kotthoff ***@***.***> wrote:

> Quoting from your report:
>
> The training set is further partitioned into 5-fold/subset
> cross-validation.
> Each fold serves as a separate training-validation split for the outer
> cross-validation loop.
>
> So you have a train/test split. Then you partition that into 5 folds. Then
> you partition each fold again into a train/test split? And that wouldn't be
> the outer resampling.
>
> Nested within the outer loop, the inner loop facilitates hyperparameter
> optimization by further
> partitioning the training folds into 3 non-overlapping folds.
>
> 3 folds? Is that in addition to the 5 above? Or should the 3 be a 5?
> Please check your report carefully.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/COSC5557/pipeline-optimization-abi-awosola/pull/2#issuecomment-2099569061>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BEAXPKM5LHSOSEDNDAWB7WLZBF6I3AVCNFSM6AAAAABGN4WOPWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAOJZGU3DSMBWGE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

abi-awosola commented:
[COSC_5557_PipelineOptmzn.pdf](https://github.com/COSC5557/pipeline-optimization-abi-awosola/files/15255822/COSC_5557_PipelineOptmzn.pdf)


I fixed this: replacing `outer_cv.split(X_train)` with `outer_cv.split(X)`, in the classification part (it looks okay in the regression part). And I updated my statement on outer cv splitting.

I do apologize for not getting all together at once.
